### ISSUE TITLE
awaitWriteFinish not taking effect on file changes

### ISSUE THREAD

**Issue by ng-dan** [paulmillr/chokidar#381]
I have the following code:
```JavaScript
var watcher = require('chokidar');
var watch = watcher.watch('file,dir or glob', {
        persistent: true
    });

watch.on('change', function (path, stats) {
    if (stats && stats.size > 0) {
        console.log("Reading file with size: " + stats.size);
    }
}
```

However when I upload a file (from WebStorm) to where it is watching it will trigger the change event multiple times as the file is being uploaded, like so (the output from one file upload):
```
Reading file with size: 8192
Reading file with size: 16384
Reading file with size: 24576
Reading file with size: 32768
Reading file with size: 40960
Reading file with size: 49152
Reading file with size: 57344
Reading file with size: 65536
Reading file with size: 73728
Reading file with size: 74094
```

I've tried the usePolling option as well as the awaitWriteFinish however it does not help to fix this issue.

Any help on the matter would be appreciated.

**Comment by es128** [comment]
This is specifically what `awaitWriteFinish` was meant to provide a solution to. Without it, this is expected behavior. 

Perhaps you need to tweak its poll interval to be even longer for this use case.

// @alexpusch

**Comment by alexpusch** [comment]
What OS? What is the method of the upload? Ftp? If the upload is slow you might need to increase the awaitWriteFinish.stabilityThreshold option

**Comment by ng-dan** [comment]
This is on a server running Debian 8, and I'm using SFTP in WebStorm to upload the file. I'll try tweaking the stabilityThreshold option and see what affect that has.

**Comment by ng-dan** [comment]
I've changed it to include the awaitWriteFinish option however it hasn't made a difference - it still does the exact same thing. Perhaps I've done something wrong in the following?

```JavaScript
var watch = watcher.watch('file,dir or glob', {
        persistent: true,
        awaitWriteFinish: {
            stabilityThreshold: 60000,
            pollInterval: 100
        }
    });
```

**Comment by es128** [comment]
If it hasn't made any difference then you most likely don't have chokidar 1.2.0 installed.

**Comment by ng-dan** [comment]
I've update chokidar to 1.2.0, but it's still doing the same thing when it comes to a file being uploaded.

**Comment by alexpusch** [comment]
How big are the uploaded files? How fast is the upload? I can't think of a reason why an SFTP upload would change anything in the process.

**Comment by ng-dan** [comment]
The files aren't big at all, around 73Kb and it uploads in less than a minute.

**Comment by alexpusch** [comment]
Could you make sure you are defiantly on version 1.2.0? check the version in node_modules/chokidar/package.json, and restart your app again just in case

**Comment by nono** [comment]
By the way, I think that `stabilityThreshold` is not really well explained in the README. It says:

> Amount of time in milliseconds for a file size to remain constant before emitting its event

But, if I look at the code, I see:

```js
// https://github.com/paulmillr/chokidar/blob/master/index.js#L262
now - this._pendingWrites[path].creationTime > threshold
```

The threshold starts at the first event for this file, not the last event that changes the file size.

@es128 @alexpusch I am right? If so, it is the README or the code that should be fixed?

I think for a case like this issue, fixing the code by checking against a last file size modification date can help.

**Comment by ng-dan** [comment]
@alexpusch I've checked the version of chokidar and it is 1.2.0. I have also restarted my app and it's still doing the same thing.

**Comment by nono** [comment]
I've created a PR (#389) to show what I mean to change in code.

**Comment by es128** [comment]
@ng-dan can you try it with the code from the #389 PR and let us know what happens?

**Comment by ng-dan** [comment]
I've tried it with the updated code from the PR (by replacing the contents of /usr/lib/node_modules/chokidar/index.js - where the module is installed) and it still isn't making a difference. Perhaps I've updated it wrong?

**Comment by es128** [comment]
Could you please:
* add a timestamp to your log messages from your original test
* add a `raw` event listener that logs out the events received from there with timestamps
* share your real code so others can try to reproduce the problem

**Comment by alexpusch** [comment]
@ng-dan - You can reference a specific commit in your package.json:

in your package.json reference chokidar as follows:
```json
  "dependencies": {
   "chokidar": "git@github.com:paulmillr/chokidar.git#88a1da0f78ebbbc9a8590312cdbaabbe4bdd0d4a"
  }
```
then npm install. Watch out of confusing globally installed packages and locally installed ones.

**Comment by ng-dan** [comment]
@alexpusch I've tried that, however I'm getting the following error:
```

npm ERR! git clone git@github.com:git@github.com:paulmillr/chokidar.git Cloning into bare repository '/root/.npm/_git-remotes/git-github-com-git-github-com-paulmillr-chokidar-git-d57e6fb8'...
npm ERR! git clone git@github.com:git@github.com:paulmillr/chokidar.git Permission denied (publickey).
npm ERR! git clone git@github.com:git@github.com:paulmillr/chokidar.git fatal: Could not read from remote repository.
npm ERR! git clone git@github.com:git@github.com:paulmillr/chokidar.git
npm ERR! git clone git@github.com:git@github.com:paulmillr/chokidar.git Please make sure you have the correct access rights
npm ERR! git clone git@github.com:git@github.com:paulmillr/chokidar.git and the repository exists.
npm ERR! notarget No compatible version found: chokidar@'git@github.com:paulmillr/chokidar.git#88a1da0f78ebbbc9a8590312cdbaabbe4bdd0d4a'
npm ERR! notarget Valid install targets:
npm ERR! notarget ["1.0.0-rc1","1.0.0-rc1.1","1.0.0-rc2","0.1.1","0.2.0","0.2.1","0.2.2","0.2.3","0.2.4","0.2.5","0.2.6","0.3.0","0.4.0","0.5.0","0.5.1","0.5.2","0.5.3","0.6.0","0.6.1","0.6.2","0.6.3","0.7.0","0.7.1","0.8.0","0.8.1","0.8.2","0.8.3","0.8.4","0.9.0","0.10.0","0.10.1","0.10.2","0.10.3","0.10.4","0.10.5","0.10.6","0.10.7","0.10.8","0.10.9","0.11.0","0.11.1","0.12.0","0.12.1","0.12.2","0.12.3","0.12.4","0.12.5","0.12.6","1.0.0-rc1","1.0.0-rc1.1","1.0.0-rc2","1.0.0-rc3","1.0.0-rc4","1.0.0-rc5","1.0.0","1.0.1","1.0.2","1.0.3","1.0.4","1.0.5","1.0.6","1.1.0","1.2.0"]
npm ERR! notarget
npm ERR! notarget This is most likely not a problem with npm itself.
npm ERR! notarget In most cases you or one of your dependencies are requesting
npm ERR! notarget a package version that doesn't exist.
```

**Comment by alexpusch** [comment]
You don't have a github private key configured. Either config one, or change the reference to "https://github.com/paulmillr/chokidar.git#88a1da0f78ebbbc9a8590312cdbaabbe4bdd0d4a"

**Comment by ng-dan** [comment]
@alexpusch I've changed the reference, but I'm now getting the following when I try to npm install:
```
npm ERR! not a package /tmp/npm-28884-EH5hD-Lq/github.com/paulmillr/chokidar.git
npm ERR! Error: ENOENT, open '/tmp/npm-28884-EH5hD-Lq/github.com/paulmillr/chokidar.git-unpack/package.json'
```

**Comment by es128** [comment]
Try installing directly from master branch with `npm i paulmillr/chokidar`

**Comment by ng-dan** [comment]
That worked. Here's the log with timestamps:
```
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] Reading file with size: 8192
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] Reading file with size: 16384
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] Reading file with size: 24576
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] Reading file with size: 32768
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] Reading file with size: 40960
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] Reading file with size: 49152
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] Reading file with size: 57344
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] Reading file with size: 65536
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] change on filesizetest.html
[Mon Nov 09 2015 09:56:32 GMT+0000 (GMT)] Reading file with size: 67380
```

And here's my real code that I'm currently using for this:
```JavaScript
var watcher = require('chokidar');

var watch = watcher.watch('file,dir or glob', {
    persistent: true,
    awaitWriteFinish: {
        stabilityThreshold: 60000,
        pollInterval: 100
    }
});

watch.on('raw', function(path, stats) {
    console.log("[" + new Date() + "] " + path + " on " + stats);
});

watch.on('change', function (path, stats) {
    if (stats && stats.size > 0) {
        console.log("[" + new Date() + "] Reading file with size: " + stats.size);
    }
});

watch.add('db');
```

**Comment by es128** [comment]
Your real code says `'file,dir or glob'` for the watch paths?

**Comment by ng-dan** [comment]
Yeah. Should it be something else instead?

**Comment by es128** [comment]
Yes it should be the path(s) you want to watch. `'db'` I suppose (and remove the `.add()` at the bottom). But then I don't get why the `path` in events isn't more like `db/filesizetest.html`

**Comment by ng-dan** [comment]
Alright, but how come it shows this example in the readme?

```JavaScript
var watcher = chokidar.watch('file, dir, glob, or array', {
  ignored: /[\/\\]\./,
  persistent: true
});
```

**Comment by es128** [comment]
Please use your `change` event handler as an `all` event handler instead, and show which type of event is being emitted. I wonder if there are `unlink` events happening in between all this because `stat` calls are failing that reset the `awaitWriteFinish` handling.

**Comment by ng-dan** [comment]
How exactly is this meant to be done? From what I've done, I can't seem to get any useful information out of it.

**Comment by nono** [comment]
```js
watch.on('all', function(what, path) {
    console.log("[" + new Date() + "] " + what + ": " + path);
});
```

**Comment by ng-dan** [comment]
Thanks @nono 

I've put that code in and this is the debug log:
```
[Tue Nov 10 2015 14:43:56 GMT+0000 (GMT)] change: db/filesizetest.html
[Tue Nov 10 2015 14:43:56 GMT+0000 (GMT)] change: db/filesizetest.html
[Tue Nov 10 2015 14:43:57 GMT+0000 (GMT)] change: db/filesizetest.html
[Tue Nov 10 2015 14:43:57 GMT+0000 (GMT)] change: db/filesizetest.html
[Tue Nov 10 2015 14:43:57 GMT+0000 (GMT)] change: db/filesizetest.html
[Tue Nov 10 2015 14:43:57 GMT+0000 (GMT)] change: db/filesizetest.html
[Tue Nov 10 2015 14:43:57 GMT+0000 (GMT)] change: db/filesizetest.html
[Tue Nov 10 2015 14:43:57 GMT+0000 (GMT)] change: db/filesizetest.html
[Tue Nov 10 2015 14:43:57 GMT+0000 (GMT)] change: db/filesizetest.html
[Tue Nov 10 2015 14:43:57 GMT+0000 (GMT)] change: db/filesizetest.html
```

**Comment by es128** [comment]
Please try with

```js
var watcher = require('chokidar');

var watch = watcher.watch('db', {
    persistent: true,
    ignoreInitial: true,
    awaitWriteFinish: {
        stabilityThreshold: 60000,
        pollInterval: 100
    }
});

watch.on('raw', function(path, stats) {
    console.log("[" + new Date() + "]  RAW: " + path + " -- " + (stats ? stats.size : ''));
});

watch.on('all', function (ev, path, stats) {
    console.log("[" + new Date() + "] " + ev + ": " + path + " -- " + (stats ? stats.size : ''));
});
```

**Comment by nono** [comment]
@es128 for raw, it's not `event, path, stats` for the callback?

**Comment by ng-dan** [comment]
```
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)] change: db/filesizetest.html -- 0
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)] change: db/filesizetest.html -- 8192
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)] change: db/filesizetest.html -- 16384
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)] change: db/filesizetest.html -- 24576
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)] change: db/filesizetest.html -- 32768
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)] change: db/filesizetest.html -- 40960
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)] change: db/filesizetest.html -- 49152
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)] change: db/filesizetest.html -- 57344
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:30 GMT+0000 (GMT)] change: db/filesizetest.html -- 65536
[Tue Nov 10 2015 15:07:31 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:31 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:31 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:31 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:31 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:31 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:31 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:31 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:31 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:31 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:07:31 GMT+0000 (GMT)] change: db/filesizetest.html -- 67381
```

**Comment by es128** [comment]
Oops @nono is right. I wrote that quickly without trying it. 

But anyway, it's showing enough information to negate my suspicion. And I'm back to being clueless about what's wrong. 

@ng-dan, please run

```
npm cache clean
npm i paulmillr/chokidar
```

Then please confirm the problem is still occurring, and post the results from `npm ls`.

**Comment by ng-dan** [comment]
Here's the output of npm ls chokidar after running those commands.
```
/usr/lib
└── chokidar@1.2.0  (git://github.com/paulmillr/chokidar#5f0df3f93e8ea86f44a728ce9bc4f3406b29e86c)
```

And here's the output from the script:
```
[Tue Nov 10 2015 15:19:59 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:19:59 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:19:59 GMT+0000 (GMT)] change: db/filesizetest.html -- 0
[Tue Nov 10 2015 15:19:59 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:19:59 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:19:59 GMT+0000 (GMT)] change: db/filesizetest.html -- 8192
[Tue Nov 10 2015 15:19:59 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:19:59 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:19:59 GMT+0000 (GMT)] change: db/filesizetest.html -- 16384
[Tue Nov 10 2015 15:19:59 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:19:59 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:19:59 GMT+0000 (GMT)] change: db/filesizetest.html -- 24576
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)] change: db/filesizetest.html -- 32768
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)] change: db/filesizetest.html -- 40960
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)] change: db/filesizetest.html -- 49152
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)] change: db/filesizetest.html -- 57344
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)] change: db/filesizetest.html -- 65536
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)]  RAW: change -- undefined
[Tue Nov 10 2015 15:20:00 GMT+0000 (GMT)] change: db/filesizetest.html -- 67381
```

**Comment by es128** [comment]
So it's quite persuasive that `awaitWriteFinish` is not doing its job in this circumstance, but we aren't much closer to figuring out why yet.

**Comment by notmodified** [comment]
I'm very new to node so I might have missed the mark here. I couldn't recreate the problem without bringing in the async writing and I have absolutely no familiarity with the codebase so I have no clue how bad a mess I have made.

Disclaimers aside I hope I help to at least narrow down the search a little.

**Comment by alexpusch** [comment]
Ok, so I think I figured what's going on. AwaitWriteFinish suppresses the 'add' event from firing until the files size stops changing. In @ng-dan case the file already exists, and he is updating it trough Web Storms deployment feature. Naturally only change events fire. It is possible to update the awaitWriteFinish feature to supress consecutive  'change' events as well but do we really want too?

@notmodified commit tries to do so, but I did not verify it works. (BTW Thanks @notmodified :) )
